internal/ctrlflow/ctrlflow.go

```diff
@@ -179,21 +193,47 @@ func Obfuscate(fset *token.FileSet, ssaPkg *ssa.Package, files []*ast.File, obfR
 			if junkCount > 0 {
 				addJunkBlocks(ssaFunc, junkCount, obfRand)
 			}
+			var dispatchers []dispatcherInfo
 			for i := 0; i < passes; i++ {
-				applyFlattening(ssaFunc, obfRand)
+				if info := applyFlattening(ssaFunc, obfRand); info != nil {
+					dispatchers = append(dispatchers, info)
+				}
 			}
 			fixBlockIndexes(ssaFunc)
+			return dispatchers
 		}
 
-		applyObfuscation(ssaFunc)
+		dispatchers := applyObfuscation(ssaFunc)
 		for _, anonFunc := range ssaFunc.AnonFuncs {
-			applyObfuscation(anonFunc)
+			dispatchers = append(dispatchers, applyObfuscation(anonFunc)...)
+		}
+
+		// Because of ssa package api limitations, implementation of hardening for control flow flattening dispatcher
+		// is implemented during converting by replacing key values with obfuscated ast expressions
+		var prologues []ast.Stmt
+		if len(dispatchers) > 0 && hardening != nil {
+			ssaRemap := make(map[ssa.Value]ast.Expr)
+			for _, dispatcher := range dispatchers {
+				decl, stmt := hardening.Apply(dispatcher, ssaRemap, obfRand)
+				if decl != nil {
+					newFile.Decls = append(newFile.Decls, decl)
+				}
+				if stmt != nil {
+					prologues = append(prologues, stmt)
+				}
+			}
+			funcConfig.SsaValueRemap = ssaRemap
+		} else {
+			funcConfig.SsaValueRemap = nil
 		}
```

Comment:
Shouldn't this be nil already? If not declare it as nil before the if and drop the else.

Reply:
No, `funcConfig` is reused in iterations and `SsaValueRemap` must be explicitly set to nil